### PR TITLE
Fix permuted sum precision issue for lower precision on CPU

### LINKED ISSUE #83149: bf16 strided tensor wrong calculation
### 🐛 Describe the bug

Issue when transpose is taken on bfloat16 tensor and passed to sum. the output is not correct. Example has been described below:
If you run same code in float32 then it works fine.
```
import torch
x = torch.ones([10, 13, 3, 3], dtype=torch.bfloat16)
x_trans = x.transpose(2, 3)
x_sum = torch.sum(x_trans, (0, 1, 2))
print(x_sum)
```

```
output: tensor([432., 432., 432.], dtype=torch.bfloat16)
but except output is tensor([390., 390., 390.])
```

### Versions

Collecting environment information...
PyTorch version: 1.12.0
Is debug build: False
CUDA used to build PyTorch: None
ROCM used to build PyTorch: N/A

OS: Ubuntu 20.04.4 LTS (x86_64)
GCC version: (Ubuntu 9.4.0-1ubuntu1~20.04.1) 9.4.0
Clang version: Could not collect
CMake version: version 3.16.3
Libc version: glibc-2.31

Python version: 3.8.10 (default, Jun 22 2022, 20:18:18)  [GCC 9.4.0] (64-bit runtime)
Is CUDA available: False
CUDA runtime version: No CUDA
GPU models and configuration: No CUDA
Nvidia driver version: No CUDA
cuDNN version: No CUDA
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

Versions of relevant libraries:
[pip3] mypy-extensions==0.4.3
[pip3] numpy==1.22.3
[pip3] pytorch-lightning==1.6.5
[pip3] torch==1.12.0
[pip3] torch-tb-profiler==0.4.0
[pip3] torchmetrics==0.9.3
[pip3] torchvision==0.13.0
[conda] mkl                       2022.0.2                 pypi_0    pypi
[conda] mkl-include               2022.0.2                 pypi_0    pypi
[conda] numpy                     1.22.3                   pypi_0    pypi
[conda] pytorch-lightning         1.6.5                    pypi_0    pypi
[conda] torch                     1.12.0          pypi_0    pypi
[conda] torch-tb-profiler         0.4.0                    pypi_0    pypi
[conda] torchmetrics              0.9.3                    pypi_0    pypi
[conda] torchvision               0.13.0          pypi_0    pypi

cc @ezyang @gchanan @zou3519 @VitalyFedyunin @frank-wei

### PR DESCRIPTION
**Pull request by CaoE**
Fixes #83149
There is a limitation of `TensorIterator` reductions:
The non-permuted input tensor will be coalesced down to a 2-d tensor by `TensorIterator` whereas the permuted case may become a >2d operation (for example, two reduced dimensions and non-reduced dim). 
Since the cpu reduction loop of `TensorIterator` only operates on two dimensions at a time, this means the intermediate sums will be truncated to lower precision.

### CODE REVIEW

File: aten/src/ATen/native/ReduceOps.cpp [hunk 1]
```diff
@@ -1156,6 +1156,25 @@ std::vector<Tensor> gradient(const Tensor& self, IntArrayRef dim, int64_t edge_o
...
 
+inline bool should_use_acc_buffer(at::TensorIterator& iter) {
+  const auto ndim = iter.ndim();
+  if (!iter.device().is_cpu()) {
+    return false;
+  }
+  if (!at::isReducedFloatingType(iter.common_dtype())) {
+    return false;
+  }
+  if (ndim < 2 || iter.noutputs() != 1) {
```
Comment (jgong5) [inline aten/src/ATen/native/ReduceOps.cpp:1167]: it happens with `ndim >= 3` so should check `ndim < 3` here?

File: aten/src/ATen/native/ReduceOps.cpp [hunk 2]
```diff
@@ -1156,6 +1156,25 @@ std::vector<Tensor> gradient(const Tensor& self, IntArrayRef dim, int64_t edge_o
...
+    return false;
+  }
+  auto out_strides = iter.strides(0);
+  for (const auto dim : c10::irange(0, 2)) {
+      if (out_strides[dim] != 0) {
+        return false;
+      }
+  }
```
Comment (jgong5) [inline aten/src/ATen/native/ReduceOps.cpp:1175]: should check that we have at least three reduction dims here?

**Review by jgong5 — COMMENTED** [review]
What's the performance impact? Do we have comparison between fp32 and bf16 with the same inputs?

File: aten/src/ATen/native/ReduceOps.cpp (hunk 1 above)
Comment (CaoE) [reply aten/src/ATen/native/ReduceOps.cpp:1167]: Thanks for the comment. Fixed.

File: aten/src/ATen/native/ReduceOps.cpp (hunk 2 above)
Comment (CaoE) [reply aten/src/ATen/native/ReduceOps.cpp:1175]: Checking `ndim` should have the same effect.

Comment (jgong5) [reply aten/src/ATen/native/ReduceOps.cpp:1175]: Not sure if I understand but you only check the first two dims here for reduction dims. Is that correct?

File: test/test_reductions.py [hunk 3]
```diff
@@ -1336,6 +1336,44 @@ def test_var_stability2(self, device):
...
+        dim_sequences = {
+            2: [0, 1],
+            3: [0, 1, 2],
+            4: [0, 1, 2, 3],
+            5: [0, 1, 2, 3, 4],
+        }
+
+        def helper(self, shape, reduce_dims, device, dtype):
+            permute_list = dim_sequences[len(shape)]
+            random.shuffle(permute_list)
```
[the remark this answers is not in the document]
Comment (CaoE) [reply test/test_reductions.py]: Use `permutations` instead of `random.shuffle`

File: aten/src/ATen/native/ReduceOps.cpp (hunk 2 above)
Comment (CaoE) [reply aten/src/ATen/native/ReduceOps.cpp:1175]: Yes, only check the first two dims. For example: 
Output stride is [x, x, x, x, ..., n] with x >= 0 and  n > 0. 
If the stride of the first two dims is 0 (we can also only check whether the second dim is 0), the second dim may accumulate intermediate sums which are generated on the first dim.

Comment (jgong5) [reply aten/src/ATen/native/ReduceOps.cpp:1175]: Got it.

**Review by jgong5 — COMMENTED** [review]
The changes LGTM now, still stamp if no performance impact.

**Review by mingfeima — APPROVED** [review]
Could you please update the root cause of the issue in the pr comment as well.

**Comment by CaoE** [comment]
@peterbell10 Could you please help review this PR ? Thanks.

**Comment by CaoE** [comment]
@mruberry Could you please help review this PR ? Thanks.

File: aten/src/ATen/native/ReduceOps.cpp [hunk 4]
```diff
@@ -1181,7 +1200,19 @@ TORCH_IMPL_FUNC(sum_out)
...
-    sum_stub(iter.device_type(), iter);
+    // Here is a limitation of TensorIterator reductions for permuted input with lower precision on CPU.
+    // Consider the case: TensorIterator coalesces such input and output to >= 2 dims tensors,
+    // and the output stride is [0, 0, x, x, ...] with x >= 0 (two reduced dimensions and non-reduced dims).
+    // Since the reduction loop only operates on two dimensions at a time,
+    // the intermediate sums is forced to do accumulation in the second reduced dim with lower precision.
+    // See https://github.com/pytorch/pytorch/issues/83149
+    if (should_use_acc_buffer(iter)) {
+      auto tmp_output = at::empty(result.sizes(), result.options().dtype(kFloat));
+      at::sum_outf(self.to(ScalarType::Float), opt_dim, keepdim, /*dtype=*/c10::nullopt, tmp_output);
```
Comment (peterbell10) [inline aten/src/ATen/native/ReduceOps.cpp:1211]: Note that in my original comment I suggested adding mixed dtype kernels with half precision as the input and float as the output, like we have on CUDA. This is okay for now though I guess. It just won't perform as well.

**Review by peterbell10 — APPROVED** [review]
